File: src/path-safety.ts
import { existsSync, realpathSync } from "node:fs";
import path from "node:path";

const RESERVED_DIRECTORIES = new Set([".git", ".deepvibe"]);

export class PathSafetyError extends Error {
  readonly code: string;

  constructor(code: string, message: string) {
    super(message);
    this.code = code;
    this.name = "PathSafetyError";
  }
}

export function resolveProjectTargetPath(rootDir: string, relativePath: string): string {
  const rootPath = realpathSync(path.resolve(rootDir));
  const normalizedRelativePath = normalizeRelativePath(relativePath);

  if (normalizedRelativePath.length === 0) {
    throw new PathSafetyError("PATH_INVALID", "Target path must not be empty.");
  }

  if (path.isAbsolute(normalizedRelativePath) || normalizedRelativePath.startsWith("//")) {
    throw new PathSafetyError("PATH_INVALID", `Absolute paths are not allowed: ${relativePath}`);
  }

  const segments = normalizedRelativePath.split("/");

  if (segments.some((segment) => segment === "..")) {
    throw new PathSafetyError("PATH_INVALID", `Path traversal is not allowed: ${relativePath}`);
  }

  if (segments.some((segment) => RESERVED_DIRECTORIES.has(segment))) {
    throw new PathSafetyError("PATH_RESERVED", `Reserved directories cannot be modified: ${relativePath}`);
  }

  const targetPath = path.resolve(rootPath, normalizedRelativePath);
  const boundaryPath = findBoundaryPath(rootPath, targetPath);
  const normalizedBoundary = normalizeForComparison(boundaryPath);
  const normalizedRoot = normalizeForComparison(rootPath);

  if (normalizedBoundary !== normalizedRoot && !normalizedBoundary.startsWith(`${normalizedRoot}${path.sep}`)) {
    throw new PathSafetyError("PATH_ESCAPE", `Resolved path escapes the project root: ${relativePath}`);
  }

  return targetPath;
}

export function resolveExistingProjectPath(rootDir: string, relativePath: string): string {
  const targetPath = resolveProjectTargetPath(rootDir, relativePath);

  if (!existsSync(targetPath)) {
    throw new PathSafetyError("PATH_MISSING", `Path does not exist: ${relativePath}`);
  }

  return realpathSync(targetPath);
}

export function normalizeRelativePath(filePath: string): string {
  return filePath.replace(/\\/gu, "/").replace(/^\.\/+/u, "");
}

function findBoundaryPath(rootPath: string, targetPath: string): string {
  if (existsSync(targetPath)) {
    return realpathSync(targetPath);
  }

  let currentPath = path.dirname(targetPath);

  while (normalizeForComparison(currentPath) !== normalizeForComparison(rootPath)) {
    if (existsSync(currentPath)) {
      return realpathSync(currentPath);
    }

    const parentPath = path.dirname(currentPath);

    if (parentPath === currentPath) {
      break;
    }

    currentPath = parentPath;
  }

  return rootPath;
}

function normalizeForComparison(targetPath: string): string {
  return path.normalize(targetPath).toLowerCase();
}

File: internal/workspace/path_safety.go
package workspace

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

var reservedDirectories = map[string]bool{
	".git":      true,
	".deepvibe": true,
}

type PathSafetyError struct {
	Code    string
	Message string
}

func (e *PathSafetyError) Error() string {
	return e.Message
}

func ResolveProjectTargetPath(rootDir string, relativePath string) (string, error) {
	rootPath, err := filepath.Abs(rootDir)
	if err != nil {
		return "", err
	}
	rootPath, err = filepath.EvalSymlinks(rootPath)
	if err != nil {
		return "", err
	}

	normalizedRelativePath := NormalizeRelativePath(relativePath)
	if normalizedRelativePath == "" {
		return "", &PathSafetyError{Code: "PATH_INVALID", Message: "Target path must not be empty."}
	}
	if filepath.IsAbs(normalizedRelativePath) || strings.HasPrefix(normalizedRelativePath, "//") {
		return "", &PathSafetyError{Code: "PATH_INVALID", Message: fmt.Sprintf("Absolute paths are not allowed: %s", relativePath)}
	}

	segments := strings.Split(normalizedRelativePath, "/")
	for _, segment := range segments {
		if segment == ".." {
			return "", &PathSafetyError{Code: "PATH_INVALID", Message: fmt.Sprintf("Path traversal is not allowed: %s", relativePath)}
		}
		if reservedDirectories[segment] {
			return "", &PathSafetyError{Code: "PATH_RESERVED", Message: fmt.Sprintf("Reserved directories cannot be modified: %s", relativePath)}
		}
	}

	targetPath := filepath.Join(rootPath, filepath.FromSlash(normalizedRelativePath))
	boundaryPath, err := findBoundaryPath(rootPath, targetPath)
	if err != nil {
		return "", err
	}

	normalizedBoundary := normalizeForComparison(boundaryPath)
	normalizedRoot := normalizeForComparison(rootPath)
	if normalizedBoundary != normalizedRoot && !strings.HasPrefix(normalizedBoundary, normalizedRoot+string(filepath.Separator)) {
		return "", &PathSafetyError{Code: "PATH_ESCAPE", Message: fmt.Sprintf("Resolved path escapes the project root: %s", relativePath)}
	}

	return targetPath, nil
}

func ResolveExistingProjectPath(rootDir string, relativePath string) (string, error) {
	targetPath, err := ResolveProjectTargetPath(rootDir, relativePath)
	if err != nil {
		return "", err
	}
	if _, err := os.Stat(targetPath); err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return "", &PathSafetyError{Code: "PATH_MISSING", Message: fmt.Sprintf("Path does not exist: %s", relativePath)}
		}
		return "", err
	}
	return filepath.EvalSymlinks(targetPath)
}

func NormalizeRelativePath(filePath string) string {
	path := filepath.ToSlash(filePath)
	return strings.TrimPrefix(path, "./")
}

func findBoundaryPath(rootPath string, targetPath string) (string, error) {
	if _, err := os.Stat(targetPath); err == nil {
		return filepath.EvalSymlinks(targetPath)
	}

	currentPath := filepath.Dir(targetPath)
	for normalizeForComparison(currentPath) != normalizeForComparison(rootPath) {
		if _, err := os.Stat(currentPath); err == nil {
			return filepath.EvalSymlinks(currentPath)
		}
		parentPath := filepath.Dir(currentPath)
		if parentPath == currentPath {
			break
		}
		currentPath = parentPath
	}

	return rootPath, nil
}

func normalizeForComparison(targetPath string) string {
	return strings.ToLower(filepath.Clean(targetPath))
}
